### DISCUSSION TITLE
when to `.await` JoinHandle type generated by `tokio::spawn`

### DISCUSSION THREAD

**Discussion by Camusugar** [tokio-rs/tokio#4257] [General] [answered]
If I comment out those three lines of code, the program exits early and does not print out the full result. What I'm asking is, is there a best practice for when I should `.await` the JoinHandle type generated by `tokio::spawn`  ? thank you!

```rust
#[tokio::main]
async fn main() {

    let mut handles = Vec::with_capacity(10);
    for i in 0..10 {
        handles.push(tokio::spawn(my_bg_task(i)));
    }
    

// 🔺🔺🔺
    for handle in handles {
        handle.await.unwrap();
    }
// 🔺🔺🔺


}

async fn my_bg_task(i: u64) {
    println!("{}",i)
}

**Comment by Darksonn** [comment, marked as answer]
When you return from `main`, the runtime will start shutting down, which includes killing all other spawned tasks. You should await tasks when you need to wait for them to finish.

**Reply by Camusugar** [reply]
thank you

**Comment by Camusugar** [comment]
I have some new understanding of this problem. This actually involves the shutdown of runtime. What happens when a runtime is dropped can be found at [this  link](https://docs.rs/tokio/1.14.0/tokio/runtime/struct.Runtime.html).
So what will be closed? Unblocked `worker thread`. And what will be waiting to be completed?
`blocking thread`（generated by `spawn_blocking`） and worker threads that are already blocked by some synchronization code（for example ,the code has `std::thread::sleep`).
But in my sample code for this problem, my_bg_task function doesn't have any code that makes the runtime wait for finish. when main finishes executing,runtime is dropped and all the unexecuted tasks are discarded.

**Reply by Darksonn** [reply]
Yes. Generally the solution is to prevent the runtime from starting shutdown until your tasks have completed.
